src/Datasource/QueryTrait.php

```diff
@@ -356,30 +356,85 @@ public function getMapReducers(): array
      * the return value for this query's result. Formatter functions are applied
      * after all the `MapReduce` routines for this query have been executed.
      *
-     * If the second argument is set to true, it will erase previous formatters
-     * and replace them with the passed first argument.
+     * Formatting callbacks will receive two arguments, the first one being an object
+     * implementing `\Cake\Collection\CollectionInterface`, that can be traversed and
+     * modified at will. The second one being the query instance on which the formatter
+     * callback is being applied.
      *
-     * ### Example:
+     * Usually the query instance received by the formatter callback is the same query
+     * instance on which the callback was attached to, except for in a joined
+     * association, in that case the callback will be invoked on the association source
+     * side query, and it will receive that query instance instead of the one on which
+     * the callback was originally attached to - see the examples below!
+     *
+     * ### Examples:
+     *
+     * Return all results from the table indexed by id:
      *
      * ```
-     * // Return all results from the table indexed by id
      * $query->select(['id', 'name'])->formatResults(function ($results) {
-     *   return $results->indexBy('id');
+     *     return $results->indexBy('id');
      * });
+     * ```
+     *
+     * Add a new column to the ResultSet:
      *
-     * // Add a new column to the ResultSet
+     * ```
      * $query->select(['name', 'birth_date'])->formatResults(function ($results) {
-     *   return $results->map(function ($row) {
-     *     $row['age'] = $row['birth_date']->diff(new DateTime)->y;
-     *     return $row;
-     *   });
+     *     return $results->map(function ($row) {
+     *         $row['age'] = $row['birth_date']->diff(new DateTime)->y;
+     *
+     *         return $row;
+     *     });
+     * });
+     * ```
+     *
+     * Add a new column to the results with respect to the query's hydration configuration:
+     *
+     * ```
+     * $query->formatResults(function ($results, $query) {
+     *     return $results->map(function ($row) {
+     *         $data = [
+     *             'bar' => 'baz',
+     *         ];
+     *
+     *         if ($query->isHydrationEnabled()) {
+     *             $row['foo'] = new Foo($data)
+     *         } else {
+     *             $row['foo'] = $data;
+     *         }
+     *
+     *         return $row;
+     *     });
+     * });
+     * ```
+     *
+     * Retaining access to the original query instance by inheriting the query variable:
+     *
+     * ```
+     * // Assuming a `Articles belongsTo Authors` association that uses the join strategy
+     *
+     * $articlesQuery->contain('Authors', function ($authorsQuery) {
+     *     return $query->formatResults(function ($results, $query) use ($authorsQuery) {
```

Comment:
```suggestion
     *     return $articlesQuery->formatResults(function ($results, $query) use ($authorsQuery) {
```

Reply:
Actually `$authorsQuery`, I hope that was more or less just a typo, and not a testament to this feature's potential to even confuse experienced users 🙃

Reply:
Just muddy brain from a long day :)